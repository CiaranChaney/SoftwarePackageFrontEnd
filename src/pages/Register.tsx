import {
  Avatar,
  Box,
  Button,
  Container,
  createTheme,
  CssBaseline,
  Grid,
  TextField,
  ThemeProvider,
  Typography,
} from "@mui/material";
import { LockOutlined } from "@mui/icons-material";
import { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import axios from "axios";
import Favicon from "../components/Favicon";

const darkTheme = createTheme({
  palette: {
    mode: "dark",
  },
});

const Register = () => {
  document.title = "Register";

  const [username, setUserName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const navigate = useNavigate();

  const validateEmail = (email) => {
    const regex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return regex.test(email);
  };

  const handleRegister = async (event) => {
    event.preventDefault();
    if (!validateEmail(email)) {
      console.log("Invalid email");
      setErrorMessage("Invalid email");
      return;
    }

    try {
      const response = await axios.post("https://ciaranchaney.com:443/register", {
        username: username,
        email: email,
        password: password,
      });
      console.log(response.data);

      if (response.data.token !== undefined) {
        const token = response.data.token;
        localStorage.setItem("token", token);
        console.log("Registered in successfully");
        navigate("/Library");
        window.location.reload();
        console.log("Redirecting to library...");
      } else {
        console.log("Invalid registration");
      }
    } catch (error) {
      console.error("Error logging in:", error);
    }
  };

  return (
    <ThemeProvider theme={darkTheme}>
      <Favicon />
      <Container maxWidth={"xs"}>
        <CssBaseline />
        <Box
          sx={{
            marginTop: 20,
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
          }}
        >
          <Avatar sx={{ m: 1, bgcolor: "primary.light" }}>
            <LockOutlined />
          </Avatar>
          <Typography variant={"h5"}>Sign up</Typography>
          <Box sx={{ mt: 3 }}>
            <form onSubmit={handleRegister}>
              {" "}
              {}
              <Grid container spacing={2}>
                <Grid item xs={12}>
                  <TextField
                    margin={"normal"}
                    required
                    fullWidth
                    id={"username"}
                    label={"Username"}
                    name={"username"}
                    autoFocus
                    value={username}
                    onChange={(e) => setUserName(e.target.value)}
                  />
                </Grid>
                <Grid item xs={12}>
                  <TextField
                    margin={"normal"}
                    required
                    fullWidth
                    id={"email"}
                    label={"Email Address"}
                    name={"email"}
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                  />
                </Grid>
                <Grid item xs={12}>
                  <TextField
                    margin={"normal"}
                    required
                    fullWidth
                    id={"password"}
                    label={"Password"}
                    name={"Password"}
                    value={password}
                    type="password"
                    onChange={(e) => setPassword(e.target.value)}
                  />
                </Grid>
                <Typography variant={"body2"} color={"error"}>
                  {errorMessage && <div>{errorMessage}</div>}
                </Typography>
                <Button
                  type="submit"
                  fullWidth
                  variant={"contained"}
                  sx={{ mt: 3, mb: 2 }}
                >
                  Sign up
                </Button>
                <Grid container justifyContent={"flex-end"}>
                  <Grid item>
                    <Link
                      to="/login"
                      style={{ color: "white", textDecoration: "none" }}
                    >
                      Already have an account? Sign in
                    </Link>
                  </Grid>
                </Grid>
              </Grid>
            </form>
          </Box>
        </Box>
      </Container>
    </ThemeProvider>
  );
};

export default Register;
