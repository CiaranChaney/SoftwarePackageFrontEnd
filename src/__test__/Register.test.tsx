import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Register from '../pages/Register';
import {BrowserRouter, MemoryRouter} from 'react-router-dom';
import axios from 'axios';
import {toast} from "react-toastify";

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => jest.fn(),
}));

jest.mock('react-toastify', () => {
    const actualToast = jest.requireActual('react-toastify');
    return {
        ...actualToast,
        toast: {
            ...actualToast.toast,
            error: jest.fn()
        }
    };
});

describe('Register Page', () => {

    beforeAll(() => {
        Object.defineProperty(window, 'location', {
            writable: true,
            value: { reload: jest.fn() }
        });

        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { asFragment } = render(
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders Register and checks for elements', () => {
        render(
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        );

        const usernameField = screen.getByRole('textbox', { name: /username/i });
        expect(usernameField).toBeInTheDocument();

        const emailField = screen.getByLabelText(/email address/i)
        expect(emailField).toBeInTheDocument();

        const passwordField = screen.getByLabelText(/password/i)
        expect(passwordField).toBeInTheDocument();

        const signUpButton = screen.getByRole('button', { name: /Sign up/i });
        expect(signUpButton).toBeInTheDocument();

    });

    it('renders Register and checks register functionality', () => {
        render(
            <BrowserRouter>
                <Register />
            </BrowserRouter>
        );

        const usernameField = screen.getByRole('textbox', { name: /username/i });
        expect(usernameField).toBeInTheDocument();

        const emailField = screen.getByLabelText(/email address/i)
        expect(emailField).toBeInTheDocument();

        const passwordField = screen.getByLabelText(/password/i)
        expect(passwordField).toBeInTheDocument();

        const signUpButton = screen.getByRole('button', { name: /Sign up/i });
        expect(signUpButton).toBeInTheDocument();

        fireEvent.change(usernameField, { target: { value: 'username' } });
        fireEvent.change(emailField, { target: { value: 'testEmail@email.com' } });
        fireEvent.change(passwordField, { target: { value: 'testPassword' } });
        fireEvent.click(signUpButton);

    });

    it('successfully registers and navigates to the library', async () => {
        const mockPost = jest.fn().mockResolvedValue({
            data: { token: 'fakeToken' }
        });

        axios.post = mockPost;

        render(
            <MemoryRouter>
                <Register />
            </MemoryRouter>
        );

        const usernameField = screen.getByRole('textbox', { name: /username/i });
        const emailField = screen.getByLabelText(/email address/i)
        const passwordField = screen.getByLabelText(/password/i)
        const signUpButton = screen.getByRole('button', { name: /Sign up/i });

        fireEvent.change(usernameField, { target: { value: 'username' } });
        fireEvent.change(emailField, { target: { value: 'testEmail@email.com' } });
        fireEvent.change(passwordField, { target: { value: 'testPassword' } });
        fireEvent.click(signUpButton);

        await waitFor(() => expect(mockPost).toHaveBeenCalledWith('https://ciaranchaney.com:443/register', {
            username: 'username',
            email: 'testEmail@email.com',
            password: 'testPassword'
        }));

        expect(localStorage.getItem('token')).toBe('fakeToken');
        expect(window.location.reload).toHaveBeenCalled();

        localStorage.clear();
    });

    it('fails register due to existing user', async () => {
        const mockPost = jest.fn().mockRejectedValue({
            response: {
                status: 500,
                data: {
                    error: "Internal Server Error" ,
                    path:  "/register",
                }
            }
        });

        axios.post = mockPost;

        const mockConsoleError = jest.spyOn(console, 'error').mockImplementation();

        render(
            <MemoryRouter>
                <Register />
            </MemoryRouter>
        );

        fireEvent.change(screen.getByRole('textbox', { name: /username/i }), { target: { value: 'wronguser' } });
        fireEvent.change(screen.getByLabelText(/email address/i), { target: { value: 'wrongemail@email.com' } });
        fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'wrongpass' } });
        fireEvent.click(screen.getByRole('button', { name: /Sign up/i }));

        await waitFor(() => expect(mockPost).toHaveBeenCalledWith('https://ciaranchaney.com:443/register', {
            username: 'wronguser',
            email: 'wrongemail@email.com',
            password: 'wrongpass'
        }));


        await waitFor(() => expect(axios.post).toHaveBeenCalled());

        expect(toast.error).toHaveBeenCalledWith("User already exists with that email or username. Please try again.");

        expect(localStorage.getItem('token')).toBeNull();

        mockConsoleError.mockRestore();
    });
});





