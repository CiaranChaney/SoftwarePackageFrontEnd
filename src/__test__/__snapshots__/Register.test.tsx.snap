// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page renders correctly 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-hordx"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1q94i3y-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LockOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2M9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9zm9 14H6V10h12zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2"
          />
        </svg>
      </div>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign up
      </h5>
      <div
        class="MuiBox-root css-178yklu"
      >
        <form>
           
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-axj3wf-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for="username"
                  id="username-label"
                >
                  Username
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-125iqkp-MuiFormLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-14t01uy-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
                    id="username"
                    name="username"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        Username *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-rb5gc9-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="email"
                  id="email-label"
                >
                  Email Address
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-125iqkp-MuiFormLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-14t01uy-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
                    id="email"
                    name="email"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Email Address *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-rb5gc9-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="password"
                  id="password-label"
                >
                  Password
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-125iqkp-MuiFormLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-14t01uy-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
                    id="password"
                    name="Password"
                    required=""
                    type="password"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Password *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-bljml1-MuiTypography-root"
            />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1vpl9xp-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Sign up
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
              >
                <a
                  href="/login"
                  style="color: white; text-decoration: none;"
                >
                  Already have an account? Sign in
                </a>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
